src/natsTopology.ts

```diff
@@ -35,33 +26,50 @@ export const runTopologyWithNats: RunTopology =
       unpack,
       loadSnapshot,
       persistSnapshot,
-      shouldResume = defShouldResume,
+      shouldResume = async (topologyId: string) => {
+        const snapshot = await loadSnapshot(topologyId)
+        // Snapshot was found and status is not completed
+        return snapshot && snapshot?.status !== 'completed'
+      },
+      getTopologyId = defGetTopologyId,
     } = fns
-    // Merge context from nats-jobs, msg, and any optional user context
-    const extendedContext = { ...context, msg, ...options?.context }
     const { debounceMs } = options || {}
     // Unpack data
-    const data = unpack(msg.data)
-    const numAttempts = msg.info.redeliveryCount
+    const { topologyId = getTopologyId(msg), ...topologyOptions }: MsgData =
+      unpack(msg.data)
+    // Get the stream data
+    const { redeliveryCount: numAttempts, stream, streamSequence } = msg.info
+    // Merge message context, msg, and topologyId
+    const extendedContext = { ...context, stream, topologyId, msg }
     debug('Num attempts %d', numAttempts)
     // Should the topology be resumed
-    const resuming = await shouldResume(msg)
+    const resuming = await shouldResume(topologyId)
     debug('Resuming %s', resuming)
-    const { emitter, promise, getSnapshot } = resuming
+    const { emitter, promise, getSnapshot, stop } = resuming
       ? // Resume topology based on uniqueness defined in loadSnapshot
-        resumeTopology(spec, await loadSnapshot(msg), {
+        resumeTopology(spec, await loadSnapshot(topologyId), {
           context: extendedContext,
         })
       : // Run topology with data from msg
-        runTopology(spec, dag, { ...options, data, context: extendedContext })
-    // Get the stream data
-    const streamData = getStreamDataFromMsg(msg)
+        runTopology(spec, dag, {
+          ...topologyOptions,
+          context: extendedContext,
+        })
+    // Propagate abort to topology
+    if (context) {
+      context.signal.addEventListener('abort', stop)
+    }
     // Persist
     const persist = (snapshot: Snapshot) => {
-      const streamSnapshot = { ...snapshot, ...streamData, numAttempts }
+      const streamSnapshot = {
+        ...snapshot,
+        stream,
+        streamSequence,
+        numAttempts,
+      }
```

Comment:
We should not assume the user wants to persist all these things. I think we should just pass the context to the save/persist fns:

```
loadSnapshot: (extendedContext) -> snapshot
persistSnapshot: (extendedContext, snapshot) -> void
```

Reply:
We want to persist all these things and this library is highly specific to our needs at this point.

Reply:
> We want to persist all these things

_You_ assume everyone will want this but ok, let's just merge this to move things forward.